### PR TITLE
BSD: upgrade support

### PR DESCRIPTION
**Pull request by goneri**
Implement the upgrade support:

- FreeBSD: using `pkg upgrade`
- NetBSD: with `pkgin`

### CODE REVIEW

File: cloudinit/distros/netbsd.py [hunk 1]
```diff
@@ -21,10 +21,18 @@ class NetBSD(cloudinit.distros.bsd.BSD):
...
     group_add_cmd_prefix = ["groupadd"]
-    pkg_cmd_install_prefix = ["pkg_add", "-U"]
-    pkg_cmd_remove_prefix = ['pkg_delete']
+
+    def __init__(self, name, cfg, paths):
+        super().__init__(name, cfg, paths)
+        if os.path.exists("/usr/pkg/bin/pkgin"):
+            self.pkg_cmd_install_prefix = ['pkgin', '-y', 'install']
+            self.pkg_cmd_remove_prefix = ['pkgin', '-y', 'remove']
+            self.pkg_cmd_remove_prefix = ['pkgin', '-y', 'update']
```
Comment (igalic) [inline cloudinit/distros/netbsd.py:31]: copy paste error

we repeat the assignment variable `self.pkg_cmd_remove_prefix` with a different value

Comment (goneri) [reply cloudinit/distros/netbsd.py]: oh! thanks.

File: cloudinit/distros/bsd.py [hunk 2]
```diff
@@ -18,6 +18,7 @@ class BSD(distros.Distro):
     group_add_cmd_prefix = []
     pkg_cmd_install_prefix = []
     pkg_cmd_remove_prefix = []
+    pkg_cmd_upgrade_prefix = []
     # There is no need to update the package cache on NetBSD and OpenBSD
     # TODO neither freebsd nor netbsd handles a command 'upgrade'
```
Comment (blackboxsw) [inline cloudinit/distros/bsd.py:23]: Can drop this comment now?
```suggestion
```

File: cloudinit/distros/netbsd.py [hunk 3]
```diff
@@ -21,10 +21,18 @@ class NetBSD(cloudinit.distros.bsd.BSD):
...
 
     ci_sudoers_fn = '/usr/pkg/etc/sudoers.d/90-cloud-init-users'
-
     group_add_cmd_prefix = ["groupadd"]
-    pkg_cmd_install_prefix = ["pkg_add", "-U"]
-    pkg_cmd_remove_prefix = ['pkg_delete']
+
+    def __init__(self, name, cfg, paths):
+        super().__init__(name, cfg, paths)
+        if os.path.exists("/usr/pkg/bin/pkgin"):
```
Comment (blackboxsw) [inline cloudinit/distros/netbsd.py:28]: Could we get a unit test that establishes that netbsd sets pkg_cmd_install_prefix and pkg_cmd_remove_prefix based on presence/absence of pkgin?

Also is the omission intentional for the else clause not setting pkg_cmd_update_prefix or pkg_cmd_upgrade_prefix?

Comment (blackboxsw) [reply cloudinit/distros/netbsd.py:28]: If this omission is intentional can you add a comment to the else clause about why pkg_cmd_upgrade/update are omitted in that case?

Comment (goneri) [reply cloudinit/distros/netbsd.py:28]: Done

File: cloudinit/distros/bsd.py (hunk 2 above)
Comment (goneri) [reply cloudinit/distros/bsd.py]: Done

**Comment by goneri** [comment]
Thanks for the review @blackboxsw.

File: cloudinit/distros/bsd.py [hunk 4]
```diff
@@ -97,6 +97,10 @@ def package_command(self, command, args=None, pkgs=None):
             if not self.pkg_cmd_update_prefix:
                 return
             cmd = self.pkg_cmd_update_prefix
+        elif command == 'upgrade':
+            if not self.pkg_cmd_upgrade_prefix:
+                return
```
Comment (blackboxsw) [inline cloudinit/distros/bsd.py:102]: Is it worth a log.debug message here that 'package upgrade' not supported on this distro?

File: tests/unittests/test_distros/test_netbsd.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import cloudinit.distros.netbsd
+
+import unittest.mock as mock
+
+
+@mock.patch("cloudinit.distros.netbsd.os")
+def test_init_with_pkgin(m_os):
```
Comment (blackboxsw) [inline tests/unittests/test_distros/test_netbsd.py:7]: We've recently added style docs to cloud-init after our switch to pytest. https://cloudinit.readthedocs.io/en/latest/topics/hacking.html#unit-testing


I think these two tests are good candidates for consolidation into a single test with pytest.mark.parameterize which will pass the expected return_value to the mocked os.path.exists. 

We only have a few examples of moving to pytest.mark.parameterize:
tests/unittests/test_render_cloudcfg.py
cloudinit/distros/tests/test_init.py
cloudinit/net/tests/test_init.py

Also we'd like to see tests in a class declaration

File: cloudinit/distros/bsd.py (hunk 4 above)
Comment (goneri) [reply cloudinit/distros/bsd.py:102]: No, it may instead confused the user.

File: tests/unittests/test_distros/test_netbsd.py (hunk 5 above)
Comment (goneri) [reply tests/unittests/test_distros/test_netbsd.py:7]: Done

File: tests/unittests/test_distros/test_netbsd.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+@pytest.mark.parametrize('with_pkgin', (True, False))
+@mock.patch("cloudinit.distros.netbsd.os")
+def test_init(m_os, with_pkgin):
+    print(with_pkgin)
+    m_os.path.exists.return_value = with_pkgin
+    cfg = {}
+
+    distro = cloudinit.distros.netbsd.NetBSD("netbsd", cfg, None)
+    expectation = ['pkgin', '-y', 'full-upgrade'] if with_pkgin else None
+    assert distro.pkg_cmd_upgrade_prefix == expectation
```
Comment (blackboxsw) [inline tests/unittests/test_distros/test_netbsd.py:16]: Excellent, can we add an assert about what if being checked by os.path.exists
```suggestion
    assert distro.pkg_cmd_upgrade_prefix == expectation
    assert [mock.call('/usr/pkg/bin/pkgin')] == m_os.path.exists.call_args_list
```

**Pull request merged**